### PR TITLE
[AIRFLOW-5057] Provide bucket name to functions in S3 Hook when none is specified

### PR DESCRIPTION
**Pull request by feluelle**
### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-5057
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

Provide bucket name to functions in S3 Hook when none is specified.

Note: The order of arguments has changed for `check_for_prefix`.
The `bucket_name` is now optional. It falls back to the `connection schema` attribute.
- refactor code
- complete docs
- fix pylint issues

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

File: airflow/hooks/S3_hook.py [hunk 1]
```diff
@@ -16,15 +16,46 @@
...
+    """
+    Function decorator that provides a bucket name taken from the connection
+    in case no bucket name has been passed to the function and, if available, also no key has been passed.
+    """
+
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        func_params = func.__code__.co_varnames
+
+        def has_arg(name):
```
Comment (potiuk) [inline airflow/hooks/S3_hook.py:46]: Nice!

**Review by potiuk — APPROVED** [review]
Nice one!

**Review by potiuk — APPROVED** [review]
Nice one!

**Comment by feluelle** [comment]
Did not expect that to be merged that fast :D (I hope I did not break backwards compatibility somewhere, dunno) The only breaking change was to the `check_for_prefix` I think.

The background behind this change is actually that, at our company, we have an S3 Bucket for **dev** and one for **live**. We set dev or live in the connection schema and depending on which server the dags are running it writes to different s3 buckets. :)

**Comment by ashb** [comment]
We probably can't merge this in to 1.10.4 though cos of the ordering change.

**Comment by feluelle** [comment]
Yes, that's why I target 1.10.5 :P but ya probably 2.0 ?

EDIT:
nvm that was the other ticket

**Comment by ashb** [comment]
2.0 unless you can support back-compat. (No breaking changes from 1.10.x to 1.10.y is my rule)

**Comment by feluelle** [comment]
Totally agree. 👍

**Comment by feluelle** [comment]
If **we** want to have it in 1.10.4 or .5 "+1" me and I will add a PR removing the "provide bucket_name" functionality for `check_for_prefix` so the order does not need to change.

**Pull request merged**